### PR TITLE
Refactor benchmarks

### PR DESCRIPTION
**Pull request by kulbachcedric**
Hi 👋,

I spend some time on the benchmarks and tried to refactor them.
I used [vega](https://vega.github.io/vega/) for visualising the performances. 
This requires the [mkdocs-charts-plugin](https://github.com/timvink/mkdocs-charts-plugin) which sometimes fails for the `livedocs` when reloading the documentation. Further, the `index.md` in the `benchmark` folder might get pretty big at some point.

As there are some drawbacks, I would like some feedback from you to see if I am on the right track. 

Best,
Cedric
image

### CODE REVIEW

**Comment by smastelini** [comment]
Cool! As soon as possible I'll clone the branch to make local tests 😀

File: benchmarks/requirements.txt
Comment (MaxHalford) [inline benchmarks/requirements.txt:1]: If we have this, we don't need the `environment.yml` file right?

**Review by MaxHalford — COMMENTED** [review]
Honestly it looks great to me in terms of code! Great job @kulbachcedric!!

I'll let @smastelini run this and review this qualitatively. I'm a bit busy at the moment :)

**Comment by smastelini** [comment]
Hi @kulbachcedric, I was able to run the branch locally. I have some minor comments concerning style, but it's pretty impressive already! I'm pretty happy and impressed with what you got :D

Here are my thoughts: 

- It would be nice to add unit values to some measurements when applicable. For example, "Time (seconds)" or "Memory (MB)." Talking about that, using megabytes for memory would be nice. The numbers on the y-axis can get pretty big on the current scale.
- I would put the performance measurements at the top (Accuracy, MAE, R2, etc.). Just an unspoken rule, I guess, but I think it makes sense from the user standpoint.
- I got issues where some content was left out (for instance, some content was unreadable), but it was quickly fixed by changing the page zoom and resetting it to 100% again. Do the charts enable navigation, like zooming and panning? That could be useful in some cases.

Aside from these minor comments, I think the PR is already in good shape.

Random quick note, @MaxHalford and @kulbachcedric, is it worth including the pre-commit actions in the benchmark folder?

**Comment by MaxHalford** [comment]
Yes I think it's a good idea to lint the benchmarks folder with pre-commit.

File: benchmarks/requirements.txt
Comment (kulbachcedric) [reply benchmarks/requirements.txt:1]: When adding the benchmarks to the pre-commits, we could even add the packages to the setup.py?

**Comment by kulbachcedric** [comment]
I added some todos before merging. 
Feel free to add some more! 

- [ ] Add measurement labels for time and memory
- [ ] show memory consumption in Mb
- [ ] reorder performance measurements (Accuracy, MAE, R2) 
- [ ] Add zoom to Vega 
- [ ] Refactor model pipelines (add deep-river models)
- [ ] Refactor data streams (add common data streams from papers)
- [ ] Add benchmarks to pre-commit
- [ ] Add model validation tests before executing benchmarks
- [ ] Add stream validation tests before executing benchmarks
- [ ] Refactor binary classification and multi classification merge
- [ ] Add anomaly detection benchmarks

File: river/evaluate/tracks.py
Comment (MaxHalford) [inline river/evaluate/tracks.py:63]: I'm not 100% sure it's a good idea to import all datasets. Some datasets are really large and will slow down the benchmarks by a ton

Comment (kulbachcedric) [reply river/evaluate/tracks.py:63]: I'm trying to finish evaluating it right now... and you're right. 
I'm still looking for a way to incorporate the benchmarks in such a way that we can map relevant data streams such as Elec and Covertype on the one hand, but also synthetic data streams with concept drifts. 

But maybe that's anoter issue.

Comment (MaxHalford) [reply river/evaluate/tracks.py:63]: You could take a sample of each dataset. Each dataset has a `take(k)` method to select the first `k` rows in the stream.

Overall I think you/we should aim for 1 hour max to evaluate all benchmarks

**Comment by kulbachcedric** [comment]
Hi,
from my side this PR is ready for merging :-). 
I will improve it further in the next iterations.
@MaxHalford Do you know how to add the benchmarks folder to the .pre-commit hook? 

Best
Cedric

**Comment by MaxHalford** [comment]
> from my side this PR is ready for merging :-).

I'll let @smastelini own this PR :)

>  @MaxHalford Do you know how to add the benchmarks folder to the .pre-commit hook?

@kulbachcedric I think you have to add `benchmarks` at the top of `.pre-commit-config.yaml`?

File: benchmarks/model_zoo/vw.py
Comment (smastelini) [inline benchmarks/model_zoo/vw.py:26]:
```suggestion
        return {True: y_pred, False: 1.0 - y_pred}

```

File: benchmarks/model_zoo/torch.py
Comment (smastelini) [inline benchmarks/model_zoo/torch.py:72]:
```suggestion
        return self.fc(out[-1, :])

```

File: benchmarks/config.py
Comment (smastelini) [inline benchmarks/config.py:198]:
```suggestion
}

```

**Review by smastelini — CHANGES_REQUESTED** [review]
We are almost there! You did a great job revamping our benchmarks @kulbachcedric 😄 

I left some minor suggestions. My only concern that blocks merging this PR is related to the failing tests.

What do you think @kulbachcedric and @MaxHalford ?

File: river/evaluate/tracks.py
Comment (kulbachcedric) [reply river/evaluate/tracks.py]: Changed ☺️

File: river/evaluate/tracks.py
Comment (kulbachcedric) [reply river/evaluate/tracks.py]: Actually not ... 😅

File: setup.py
Comment (smastelini) [inline setup.py:106]: We need to keep in mind the latest changes in naming to deep-river :D

**Review by smastelini — APPROVED** [review]
If nothing is blocking it, let's merge this one! Thank you, once again, @kulbachcedric for the hard work! This complete revamp is a game-changer :D

**Comment by kulbachcedric** [comment]
> If nothing is blocking it, let's merge this one! Thank you, once again, @kulbachcedric for the hard work! This complete revamp is a game-changer :D

Thanks!

**Pull request merged**
